### PR TITLE
FileManager: Show information about image, audio, and font files in Properties window

### PR DESCRIPTION
**Pull request by AtkinsSJ**
The "General" tab on this window felt lonely, and it's nice to be able to view some basic information about a file, beyond its size, modification date, and that it is a file. So, this PR adds some tabs for Image, Audio, and Font files to give more information.

There are a few random other changes in related libraries too.

image

image

## NOTE: The font preview shown has since been removed, so ignore that part. :^)

image

My concern with this is, it means decoding these files inside FileManager, which is one of the more dangerous applications to become compromised since it can see the whole filesystem that the user can. Also it does slow down opening the properties window a bit. Maybe in future it could lazy-load the files when you open the file-type-specific tab.

### CODE REVIEW

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:246]: Note (to self): This display, like all runtime displays, break when the duration is 0 or very large, which can happen on files that don't specify their runtime, which is legal in FLAC (even in the streamable subset).

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:248]: Can you leave a FIXME for other channel amounts? We do report all channel amounts of all supported formats (especially QOA and FLAC), even though we downmix them to Stereo.

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:254]: All metadata formats I know of (or have implemented) don't restrict the track number to an actual decimal number. I'd rather not parse it here, but display it verbatim.

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (kleinesfilmroellchen) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:251]: Can you use the long artist output instead which will contain all artists? Have a look at how file and SoundPlayer do that.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
As requested, I've looked at the audio-related commits. I think in the future we want to be able to display *all* of the audio metadata, but this is an excellent start for now.

**Comment by AtkinsSJ** [comment]
(Push just to rebase and fix conflict, no changes yet.)

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (AtkinsSJ) [reply Userland/Applications/FileManager/PropertiesWindow.cpp:254]: This code doesn't parse it, `MetaData::track_number` is an `Optional<unsigned>` so this just turns it into a string.

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (AtkinsSJ) [reply Userland/Applications/FileManager/PropertiesWindow.cpp]: Sure!

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (nico) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:435]: Is `records()` guaranteed to be non-empty?

…looking through `MultiLocalizedUnicodeTagData::from_bytes()`, nothing seems to enforce `number_of_records > 0`, so should probably handle having no records here (?)

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (nico) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:438]: Eh, not sure what we'd do with it.

The spec says (https://color.org/specification/ICC.1-2001-04.pdf, "6.5.17 textDescriptionType"):

"The 7-bit ASCII description is to be an invariant, nonlocalizable name for consistent reference. It is preferred that both the Unicode and ScriptCode structures be properly localized."

Since serenity isn't localized, maybe we should prefer `ascii_description()` anyways?

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (nico) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:417]: This should probably move to LibGfx/ICC/SomeFile.h at some point, seems generally useful (…but we can move it once we need it in a 2nd place, having it here works for me for now).

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (nico) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:436]: nit: no `else` after `return`

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (nico) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:440]: nit: no `else` after `return`

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (nico) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:378]: Can you file an issue for this?

I noticed this too at some point while writing tests (in cee1f9b I think – if the first two lines in expect_single_frame_of_size() are swapped, jpeg tests fail, and they really shouldn't). I should've made an issue for that back then.

**Review by nico — APPROVED** [review]
The ICC bits look good to me (some commentary below, but feel free to ignore as much of it as you want).

Loading this in-process seems a bit fishy, but it's "only" metadata, so maybe it's fine. It'd be kind of nicer to ask the image service for the metadata though, maybe.

And computing the data lazily on switch to the tab with the data would be nice as well, to not slow down showing the dialog with the main tab. (Most of the time, people won't switch tabs.)

Or maybe lazily kick off loading the data once the dialog is visible, so that the data is ready once tabs are switched.

But how much doing that before merging this code is a prerequisite for merging is up to maintainers, not me, of course :)

**Comment by nico** [comment]
:bikeshed: Having a font preview seems a bit inconsistent with the other tabs. The other tabs show metadata, while the font preview shows actual data. That's kind of like showing a thumbnail of a bitmap. I'd omit the font preview.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Comment by AtkinsSJ** [comment]
Changes:
- Rebase to fix up the MimeData changes.
- Remove font preview. (You're right Nico, it isn't consistent with showing "properties" of a file. Though, we don't actually have any way of previewing a non-installed font, besides FontEditor. :thonk:)
- Move the "string from ICC tag data" code to `Profile::tag_string_data(TagSignature signature)`. It now returns `Optional` instead of `ErrorOr`, because 1) it never does anything fallible :man_facepalming: and 2) that matches `Profile::tag_data()`.
  - Make it get ASCII data from `TextDescriptionTagData`.
  - Make it handle an empty set of records in a `MultiLocalizedUnicodeTagData`.

Lazy-loading, and getting metadata from out-of-process, I'll leave for another time, but I will try and remember to make those changes. :sweat_smile: 

However, it probably makes sense to wait for #19957 to get merged first so I'll make this a draft for a bit.

**Comment by AtkinsSJ** [comment]
Rebased to match the MimeData changes.

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]
Found two small things while looking at the audio stuff again

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (LucasChollet) [inline Userland/Applications/FileManager/PropertiesWindow.cpp:376]: That's fixed now!

Comment (AtkinsSJ) [reply Userland/Applications/FileManager/PropertiesWindow.cpp:376]: Oh yeah! Thanks

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (AtkinsSJ) [reply Userland/Applications/FileManager/PropertiesWindow.cpp:258]: There's an `else` branch below, which does that. :^)

File: Userland/Applications/FileManager/PropertiesWindow.cpp
Comment (AtkinsSJ) [reply Userland/Applications/FileManager/PropertiesWindow.cpp:251]: Ooh, yeah. I think I copied this from somewhere, I'll see about fixing that too.

**Review by kleinesfilmroellchen — APPROVED** [review]

**Pull request merged**
